### ISSUE TITLE
How to describe polymorphic endpoint

### ISSUE THREAD

**Issue by raderio** [swagger-api/swagger-core#2340]
We use Spring+Jackson(Java). And in our API we can send different object to same endpoint.
For example
```
  @JsonTypeInfo(
      use = JsonTypeInfo.Id.NAME, 
      include = As.PROPERTY, 
      property = "type")
    @JsonSubTypes({
        @JsonSubTypes.Type(value = Dog.class, name = "dog"),
        @JsonSubTypes.Type(value = Cat.class, name = "cat")
    })
    public static class Animal {
    }
 
    @JsonTypeName("dog")
    public static class Dog extends Animal {
        public double barkVolume;
    }
 
    @JsonTypeName("cat")
    public static class Cat extends Animal {
        boolean likesCream;
        public int lives;
    }

@Controller
public class MyController {

    //REST service
    @RequestMapping( value = "test")
    public  @ResponseBody String save(@RequestBody  Animal animal){
        System.out.println(animal.getClass());
        return success;
    }
}
```
Sending to /test
```
{
    "type": "dog",
    "barkVolume": 23.3
}
```
will show Dog.class

Sending to /test
```
{
    "type": "cat",
    "likesCream": true,
    "lives": 42
}
```
will show Cat.class

How to describe polymorphic endpoint in Swagger?

**Comment by Martin-Wegner** [comment]
One problem is, that Swagger does not evaluate the `@JsonTypeInfo` annotation like mentioned here: https://github.com/swagger-api/swagger-core/issues/280#issuecomment-194393431 The `type` property is not a bean property and therefore Swagger does not recognize it :(

**Comment by mudit-sethi** [comment]
@frantuma Thanks for the fix. We really wanted this to be fixed. How can I try and test this fix? We are using swagger jaxrs2 plugin 2.0.8 at the moment.

**Comment by frantuma** [comment]
This is available in latest `2.0.9-SNAPSHOT`, just reference this version (in case enabling snapshot repos in your pom)

**Comment by mudit-sethi** [comment]
@frantuma I got a chance to try your fix with the following:
`@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "type")
@JsonSubTypes({
        @JsonSubTypes.Type(name = "DOG", value = DogDto.class),
        @JsonSubTypes.Type(name = "CAT", value = CatDto.class)})`

the swagger doc shows me the 'type' field but not include the sub types in it as part of the generated doc

**Comment by frantuma** [comment]
your scenario seems to be working on e.g. in [this test](https://github.com/swagger-api/swagger-core/blob/43557c57558f679668f9762f4b119acd010499f6/modules/swagger-jaxrs2/src/test/java/io/swagger/v3/jaxrs2/ReaderTest.java#L998)

can you share more details about the annotated resources and classes, the swagger configuration, and the resolved spec?

**Comment by mudit-sethi** [comment]
I have @Schema(type="object", discriminatorProperty = "type", required = true, oneOf = {Dog.class, Cat.class}) on the Animal class. I think it might be an issue with swagger-ui example not showing it correctly, when I look at the schema in UI then seems to be showing fine

**Comment by frantuma** [comment]
If i get it right, it might be an issue with swagger-ui? if this is the case, please open an issue [there](https://github.com/swagger-api/swagger-ui) (providing the spec causing the issue)

**Comment by SabhtarshaMojo** [comment]
Hit a NPE when my model had some required fields but did not have the discriminatorProperty as part of the class definition.
Could fix it by changing the code here: https://github.com/swagger-api/swagger-core/blob/6c702f1e5d5842d1abdcf5ba664f54ca615cd202/modules/swagger-core/src/main/java/io/swagger/v3/core/jackson/ModelResolver.java#L1643

Changed the condition to following and it fixed it, 
```if (modelToUpdate.getRequired() == null || !modelToUpdate.getRequired().contains(typeInfoProp)) {```

**Comment by frantuma** [comment]
thanks @SabhtarshaMojo , it was indeed a bug, fixed in #3293
